Try to execute flyweight tasks on the master

daniel-beck opened the pull request:
Try to accomplish the same goal as #3534, just differently. Do we think this approach makes sense and should be pursued further?

100% Untested.

Also add a TODO to a weird part of code called from the method I change. Does this look weird to anyone else? Originally added in https://github.com/jenkinsci/jenkins/commit/55bf883290273ac8caacf272e684151676666f42

### Proposed changelog entries

* Try to always execute flyweight tasks, such as the main builds for Pipeline or Matrix jobs, on the master.

<!-- Comment: 
The changelogs will be integrated by the core maintainers after the merge.  See the changelog examples here: https://jenkins.io/changelog/ -->

### Submitter checklist

- [n/a] JIRA issue is well described
- [x] Changelog entry appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
      * Use the `Internal: ` prefix if the change has no user-visible impact (API, test frameworks, etc.)
- [ ] Appropriate autotests or explanation to why this change has no tests
- [n/a] For dependency updates: links to external changelogs and, if possible, full diffs

daniel-beck committed:
Try to execute flyweight tasks on the master

jglick reviewed (approved):
I guess? Is there some way to test this?

daniel-beck commented:
> Is there some way to test this?

Probably really annoying to write tests for, so wanted to get the "no way" review responses out of the way first 😬

Jimilian commented:
Calculating `ConsistentHash` for all nodes on a big cluster is a real performance issue, this short-cut definitely makes sense. I thought I addressed it in #3157, but looks like that some cases were not covered (if @KostyaSha still observes `ConsistentHash` in profiler).

KostyaSha reviewed (commented):
LGTM

jglick commented:
> Probably really annoying to write tests for

Possibly not. Just need a mock task with or without a label, an online agent but zero master executors (did I get the use case right?), and have the task record which node it “ran on”.

> Try to always execute flyweight tasks, such as the main builds for Pipeline or Matrix jobs, on the master.

I think you can remove Pipeline from mention here, as it [explicitly requests the master node](https://github.com/jenkinsci/workflow-job-plugin/blob/846c3a359c04d3793dc0151ebd22f270a155949c/src/main/java/org/jenkinsci/plugins/workflow/job/WorkflowJob.java#L407). Definitely needs to be tested with matrix projects somehow.

oleg-nenashev commented on core/src/main/java/hudson/model/Node.java:
```diff
@@ -390,6 +390,7 @@ public CauseOfBlockage canTake(Queue.BuildableItem item) {
             // flyweight tasks need to get executed somewhere, if every node
             if (!(item.task instanceof Queue.FlyweightTask && (
                     this instanceof Jenkins
+                            // TODO Why is the next operator a '||' instead of a '&&'?
```
```suggestion
                            // TODO(daniel-beck): Why is the next operator a '||' instead of a '&&'?
```

daniel-beck replied:
`git blame` does this.

oleg-nenashev reviewed (approved):
The current approach looks good to me, and it is definitely more clean than https://github.com/jenkinsci/jenkins/pull/3534 . +1 for merging.

Would be also nice to have a JIRA issue, because it would be be interesting to backport it into the previous LTS line

varyvol commented:
@daniel-beck is this still a work in progress or is it ready to be reviewed? It's conflicted by the way, could you please fix it?

@Jimilian your review is pending, could you please dedicate some time to it? Thanks!

daniel-beck committed:
Fix offline behavior, add tests

daniel-beck commented:
Updated and added a rudimentary test.

Perhaps @jglick can address my confusion around `#isAcceptingTasks` being the wrong choice here? It returns true even for offline nodes (and appears to be ignored in the code further down).

jglick commented:
> Perhaps @jglick can address my confusion…

No idea offhand, sorry.

daniel-beck committed:
Remove obsolete TODOs, also check #isAcceptingTasks

oleg-nenashev reviewed (approved):
Assuming that the build passes, +1 from me. One concern is that the flyweight tasks will still go to agents if the master is offline, e.g. due to the Disk space monitor. 

My proposal is to actually assume that Flyweight tasks run on master unless explicitly stated otherwise by a task label. It is a more drastic change, and it needs wider discussion in a follow-up.

jtnord reviewed (approved):
I think my concern here would be that there can exist some flyweight tasks that are not "flyweight" but "heavyweight", and this could end up overloading the master.

In principle I don;t see anything wrong with this, but I am wondering if a node (Jenkins in this case) should be able to limit the number of flyweight tasks similar to what it does for normal Executors.

That said if the tasks truly are flyweight this should just be a problem of memory (for whatever the flyweight needs to run and it's cpu should be teeny tiny) so this is arguably be a bug in the flyweight task (but it could make jenkins look bad - but then any poorly written plugin can do silly things today anyway).

daniel-beck commented:
> I think my concern here would be that there can exist some flyweight tasks that are not "flyweight" but "heavyweight", and this could end up overloading the master.

"Heavyweight flyweight tasks" should be able to get a label assigned, then this does not happen.

Wadeck reviewed (approved):
Code looking good, nothing obviously wrong, but also I am not super familiar with this area :)

oleg-nenashev commented:
I plan to merge it tomorrow if no negative feedback
